### PR TITLE
Use builtin generics and PEP 604 for type annotations wherever possible

### PR DESCRIPTION
**Pull request by AlexWaygood**
I'm not sure if this is something we want to do or not -- but if it is something we want to do, this is what it looks like!

This PR was made using the following script, which relies on `com2ann`, `pyupgrade`, `autoflake`, `pycln`, `black` and `isort` all being installed:

<details>

```python
import subprocess
from pathlib import Path

DIRECTORIES = ("mypy", "mypyc", "scripts", "misc", "docs")
BRANCH_NAME = "future-annotations"
BAD_TEST_FILE = "mypy/test/testinfer.py"


def run(*args, check: bool = True) -> None:
    subprocess.run(list(args), check=check)


def commit(message: str) -> None:
    run("git", "add", *DIRECTORIES)
    run("git", "commit", "-m", message)


def checkout_and_sync_with_master() -> None:
    run("git", "checkout", "master")
    run("git", "pull", "upstream", "master")
    run("git", "push")


def cleanup() -> None:
    run("git", "restore", "--staged", *DIRECTORIES)
    checkout_and_sync_with_master()


def main() -> None:
    checkout_and_sync_with_master()
    run("git", "branch", "-D", BRANCH_NAME, check=False)
    run("git", "checkout", "-b", BRANCH_NAME)

    for directory in DIRECTORIES:
        run("com2ann", directory)
    commit("Convert type comments to annotations using com2ann")

    for directory in DIRECTORIES:
        run("pyupgrade", *Path(directory).rglob("*.py"), check=False)
    commit("Apply pyupgrade")

    with open(BAD_TEST_FILE) as f:
        old = f.read()
    new = old.replace(
        "callee_kinds_: list[ArgKind | tuple[ArgKind, str]],",
        "callee_kinds_: list[ArgKind | Tuple[ArgKind, str]],",
    )
    with open(BAD_TEST_FILE, "w") as f:
        f.write(new)
    commit("Fix name collision in mypy/test/testinfer.py")

    for directory in DIRECTORIES:
        run("autoflake", "--in-place", *Path(directory).rglob("*.py"))
        run("pycln", "--all", "--exclude", r"mypyc/primitives/registry\.py", directory)
    commit("Remove unused imports using autoflake and pycln")

    run("black", ".")
    run("isort", ".")
    commit("black and isort")

    run("git", "push", "--set-upstream", "origin", BRANCH_NAME, "--force")


if __name__ == "__main__":
    try:
        main()
    finally:
        cleanup()
```

</details>

### CODE REVIEW

File: misc/actions_stubs.py [hunk 1]
```diff
@@ -20,7 +20,7 @@ def apply_all(
     directory: str,
     extension: str,
     to_extension: str = "",
-    exclude: Tuple[str] = ("",),
+    exclude: tuple[str] = ("",),
```
Comment (JelleZijlstra) [inline misc/actions_stubs.py:23]: Not related to your change but pretty sure this should be `tuple[str, ...]`.

Comment (AlexWaygood) [reply misc/actions_stubs.py:23]: That looks highly plausible :) would rather not make that change in this PR, though

File: misc/incremental_checker.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ def delete_folder(folder_path: str) -> None:
         shutil.rmtree(folder_path)
 
 
-def execute(command: List[str], fail_on_error: bool = True) -> Tuple[str, str, int]:
+def execute(command: list[str], fail_on_error: bool = True) -> tuple[str, str, int]:
```
Comment (ethanhs) [inline misc/incremental_checker.py:69]: The tools seem not to have caught type comments, can we change those too?

Comment (AlexWaygood) [reply misc/incremental_checker.py:69]: Okay, I've added an additional step at the beginning, using `com2ann` to autoconvert type comments to PEP-526 annotations. It wasn't able to autoconvert them all, but I'd prefer to leave the remaining ones to a separate PR; some of them look tricky-ish.

Comment (ethanhs) [reply misc/incremental_checker.py:69]: Sounds good!

**Review by ethanhs — APPROVED** [review]

**Review by JelleZijlstra — APPROVED** [review]

**Comment by AlexWaygood** [comment]
Thanks everybody! 🥳

**Pull request merged**